src/cim/components/InfraEnv/InfraEnvAgentTable.tsx

```diff
@@ -42,30 +95,92 @@ const InfraEnvAgentTable: React.FC<InfraEnvAgentTableProps> = ({
           agents,
           bareMetalHosts,
           onEditHostname: actions.onEditHost,
-          onApprove: actions.onApprove,
+          onApprove,
         }),
         (!hideClusterColumn && clusterColumn(agents, getClusterDeploymentLink)) as TableRow<Host>,
         discoveredAtColumn,
         cpuCoresColumn,
         memoryColumn,
         disksColumn,
       ].filter(Boolean),
-    [agents, actions, getClusterDeploymentLink, hostActions, bareMetalHosts, hideClusterColumn],
+    [
+      agents,
+      actions,
+      getClusterDeploymentLink,
+      hostActions,
+      bareMetalHosts,
+      hideClusterColumn,
+      onApprove,
+    ],
   );
+
+  const hostIDs = hosts.map((h) => h.id);
+  const selectedAgents = agents.filter((a) => selectedHostIDs.includes(a.metadata?.uid || ''));
+
+  const massActions = [
+    <ChangeHostnameAction
+      key="hostname"
+      onChangeHostname={() => setMassChangeHostOpen(!isMassChangeHostOpen)}
+    />,
+  ];
+  if (onApprove) {
+    massActions.push(
+      <MassApproveAction
+        key="approve"
+        onApprove={() => setMassApproveOpen(!isMassApproveOpen)}
+        selectedAgents={selectedAgents}
+      />,
+    );
+  }
+
   return (
     <>
+      <HostToolbar
+        hostIDs={hostIDs}
+        selectedHostIDs={selectedHostIDs}
+        onSelectNone={() => setSelectedHostIDs([])}
+        onSelectAll={() => setSelectedHostIDs(hostIDs)}
+        actionItems={massActions}
+      />
       <HostsTable
         hosts={hosts}
         content={content}
         actionResolver={actionResolver}
         className={className}
+        selectedIDs={selectedHostIDs}
+        setSelectedHostIDs={setSelectedHostIDs}
+        onSelect={onSelect}
+        ExpandComponent={DefaultExpandComponent}
       >
         <HostsTableEmptyState setDiscoveryHintModalOpen={setDiscoveryHintModalOpen} />
       </HostsTable>
       <DiscoveryTroubleshootingModal
         isOpen={isDiscoveryHintModalOpen}
         setDiscoveryHintModalOpen={setDiscoveryHintModalOpen}
       />
+      <MassChangeHostnameModal
+        isOpen={isMassChangeHostOpen}
+        hosts={hosts}
+        selectedHostIDs={selectedHostIDs}
+        onChangeHostname={async (host, hostname) => {
```

Comment:
nit: It would be better to declare the handler upwards.

Reply:
fixed